sdks/python/apache_beam/runners/portability/portable_metrics.py

```diff
@@ -0,0 +1,74 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+from apache_beam.metrics import monitoring_infos
+from apache_beam.metrics.execution import MetricKey
+from apache_beam.metrics.metric import MetricName
+
+
+class ParseMonitoringInfoMixin(object):
+  @staticmethod
+  def from_monitoring_infos(monitoring_info_list, user_metrics_only=False):
+    """Groups MonitoringInfo objects into counters, distributions and gauges.
+
+    Args:
+      monitoring_info_list: An iterable of MonitoringInfo objects.
+      user_metrics_only: If true, includes user metrics only.
+    """
+    counters = {}
+    distributions = {}
+    gauges = {}
+
+    for mi in monitoring_info_list:
+      if (user_metrics_only and
+          not monitoring_infos.is_user_monitoring_info(mi)):
+        continue
+
+      key = ParseMonitoringInfoMixin._create_metric_key(mi)
+      metric_result = (monitoring_infos.extract_metric_result_map_value(mi))
+
+      if monitoring_infos.is_counter(mi):
+        counters[key] = metric_result
+      elif monitoring_infos.is_distribution(mi):
+        distributions[key] = metric_result
+      elif monitoring_infos.is_gauge(mi):
+        gauges[key] = metric_result
+
+    return counters, distributions, gauges
+
+  @staticmethod
+  def _create_metric_key(monitoring_info):
+    step_name = ParseMonitoringInfoMixin._get_step_name(monitoring_info)
+    namespace, name = monitoring_infos.parse_namespace_and_name(monitoring_info)
+    return MetricKey(step_name, MetricName(namespace, name))
+
+  @staticmethod
+  def _get_step_name(monitoring_info):
+    keys_to_check = [monitoring_infos.PTRANSFORM_LABEL,
```

Comment:
This doesn't seem correct. Why would you consider the step name to be under labels other than PTRANSFORM_LABEL

see montioring_info specs defined here
https://github.com/apache/beam/blob/d4afbabf38a3ab557625c9c091ed5f06ca6731ce/model/pipeline/src/main/proto/metrics.proto

PTRANSFORM_LABEL is the only one used for this purpose

Reply:
No, the intention is that URN + Labels defines the metric instance.

Think of URN as the class of metric, and URN+Labels defines the object instance, as an analogy.

I don't quite remember exactly what MetricKey contained, but generally our collection objects for metrics need to account for URN+Labels to correctly identify the metric instance.

MetricResult was originally designed just for user metrics, which did not have labels. Just a name and namespace. The labels concept was introduced to MonitoringInfos later. Then the name and namespace were reworked to be labels.
See the user metric MonitoringInfoSpec, defining what should be populated on an MI
https://github.com/apache/beam/blob/d4afbabf38a3ab557625c9c091ed5f06ca6731ce/model/pipeline/src/main/proto/metrics.proto#L67

Reply:
In Dataflow runner we have to translate SDK PTransform and PCollection names [back to user defined names](https://github.com/apache/beam/blob/19799cdb1ba0c9b92544a470274d1c5c3fa6ff70/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/ElementCountMonitoringInfoToCounterUpdateTransformer.java#L100). I'm still checking the code to see if that is required here as well.

Reply:
I see your point now.
I think it might be too difficult (if possible at all) to implement such a mapping right here. Since this is a portable runner, such a logic would have to be aware of all possible runner's implementation details.

I'm going to stay with checking PTRANSFORM_LABEL only for now.

Reply:
If you want to aggregate away the other labels, you could use this approach if you sum everything with the same ptransform_label